File: services/client/src/safe_socket/safe_socket.go
package safe_socket

import "io"

//TODO: Complete with a short-read/short-write tolerant implementation

func SendAll(socket io.Writer, bytes []byte) error {
	totalWritten := 0
	bytesToWrite := len(bytes)

	for totalWritten < bytesToWrite {
		n, err := socket.Write(bytes[totalWritten:])
		if err != nil {
			return err
		}
		totalWritten += n
	}
	
	return nil
}

func RecvAll(socket io.Reader, size int) ([]byte, error) {
	totalRead := 0
	buff := make([]byte, size)

	for totalRead < size {
		n, err := socket.Read(buff[totalRead:])
		if err != nil {
			return nil, err
		}
		totalRead += n
	}
	
	return buff, nil

}

File: services/server/src/safe_socket/safe_socket.py
import socket

# TODO: Complete with a short-read/short-write tolerant implementation


def recv_all(socket: socket.socket, size: int):
    buffer = bytearray()
    while len(buffer) < size:
        packet = socket.recv(size - len(buffer))
        if not packet:
            if len(buffer) == 0:
                return None
            raise RuntimeError("Error: La conexión se cortó a la mitad del mensaje")
        buffer.extend(packet)
    return bytes(buffer)


def send_all(socket: socket.socket, data: bytes):
    total_sent = 0
    while total_sent < len(data):
        sent = socket.send(data[total_sent:])
        total_sent += sent